File: backend-app/src/main/java/com/backend_app/backend_app/service/BeneficioService.java
package com.backend_app.backend_app.service;

import com.backend_app.backend_app.domain.Beneficio;
import com.backend_app.backend_app.dto.BeneficioDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;

public interface BeneficioService {

    List<BeneficioDTO> listadoBeneficios();

    Optional<BeneficioDTO> buscarBeneficioPorId(long idBeneficio);

    public BeneficioDTO convertirADTO(Beneficio beneficio);

    void insertarBeneficio(BeneficioDTO beneficioDTO);

    void actualizarBeneficio(BeneficioDTO beneficioDTO);

    void eliminarBeneficio(BeneficioDTO beneficioDTO);

    Page<BeneficioDTO> findAll(Pageable pageable);

    List<BeneficioDTO> obtenerBeneficiosAsociados();

    List<BeneficioDTO> buscarBeneficioPorNombre(String nombreBeneficio);

    List<BeneficioDTO> listadoBeneficiosCompletos();

}

File: frontend-app/src/app/services/beneficio.service.ts
import { HttpClient, HttpParams } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { Beneficio } from "../domain/Beneficio";


@Injectable({
    providedIn: 'root'
})
export class BeneficioService{

    private url: string = 'http://localhost:8080/api/beneficios';

    constructor(private http: HttpClient){

    }

    listadoBeneficios(): Observable<Beneficio[]>{
        return this.http.get<Beneficio[]>(this.url);
    }
    
    buscarBeneficioPorId(id: number): Observable<Beneficio>{
        return this.http.get<Beneficio>(`${this.url}/${id}`);
    }

    guardarActividad(beneficio: Beneficio){
        return this.http.post('http://localhost:8080/api/beneficios/guardar', beneficio);
    }
        
    editarBeneficio(beneficio: Beneficio) {
        return this.http.put(`${this.url}/actualizar`, beneficio);
    }
        
    eliminarBeneficio(idBeneficio: number){
        return this.http.delete(`${this.url}/${idBeneficio}`);
    }

    listadoBeneficiosPaginacion(page: number): Observable<any>{
        return this.http.get<any[]>(`${this.url}/page/${page}`);
    }

    listadoBeneficiosAsociados(): Observable<Beneficio[]>{
        return this.http.get<Beneficio[]>(`${this.url}/beneficios-asociados`);
    }

    buscarBeneficioPorNombre(nombreBeneficio: string): Observable<Beneficio[]>{
        const params = new HttpParams().set('nombreBeneficio', nombreBeneficio);
                    
        return this.http.get<Beneficio[]>(`${this.url}/buscar`, { params });
    }

    listadoBeneficiosCompletos(): Observable<Beneficio[]>{
        return this.http.get<Beneficio[]>(`${this.url}/completos`);
    }

}